lib/improver/database/spotdata_database.py

```diff
@@ -0,0 +1,276 @@
+                    if dim in df.columns:
+                        continue
+
+                    if dim in [coord.standard_name or coord.long_name for coord in
+                                                  c.dim_coords + c.aux_coords]:
+                        coord = c.coord(dim)
+                        col_name = col or dim
+                        if len(coord.points) == 1:
+                            col_data = coord.points[0]
+                        else:
+                            col_data = coord.points
+                    else:
+                        # Should have a conditional
+                        col_name = col or dim
+                        col_data = cube.name()
+
+                    df.insert(1, col_name, col_data)
+                    if dim != self.pivot_dim:
+                        # This is rather expensive
+                        df.set_index(col_name, append=True, inplace=True)
+                try:
+                    self.df = self.df.combine_first(df)
+                except AttributeError:
+                    self.df = df
+
+        #self.df = self.df.pivot(columns = 'forecast_period', values='vals')
+
+    def constrained_to_dataframe():
+        """
+        Turn the input cubes into a dataframe sorted by validity_time
+
+        """
+        cubes = self.cubelist
+        rows = self.primary_dim
+        coords = [coord.standard_name or coord.long_name for coord in
+                                                    self.cubelist[0].coords()]
+        for cube in cubes:
+            ignored_coord_dims = [item for item in ignored_coords if item not in cube.dim_coords]
+            print cube
+            for c in cube.slices_over('time'):
+                df = iris.pandas.as_data_frame(c, copy=False)
+
+        primary_key_iterator = self.determine_range(rows)
+        for row_val in sorted(primary_key_iterator, key=lambda x: x.points):
+            print row_val
+            records = []
+
+            # Constrain the cubes by primary_dim
+            constraint = iris.Constraint(**{rows: row_val.points})
+            selection = self.cubelist.extract(constraint)
+
+            record = dict()
+            for cube in selection:
+                df = iris.pandas.as_data_frame(cube, copy=False)
+            return
+            for iterator in iris.iterate.izip(*selection):
+                df = iris.pandas.as_data_frame(cube, copy=False)
+                for cube in iterator:
+                    dictionary = cube.coords()
+                    record = {coord.standard_name or
+                              coord.long_name: coord.points[0]
+                              for coord in [cube.coord(c) for c in cols]}
+                    records.append(record)
+            new_df = DataFrame.from_dict(records)
+            self.df = pd.concat([self.df, new_df])
+        return self.df
+
+    def create_table(self, outfile, table='test'):
```

Comment:
You're going to have to explain this function and the one below to me again. This code is probably the least accessible to members of the team so the docstrings and comments are going to be really important. It might be worth someone with more knowledge to have a look at these functions. I think we can come up with unit tests, as we do test output elsewhere in the code and I can point you to some examples, but we will have to discuss it.

Reply:
I will add some detailed comment here.